### DISCUSSION TITLE
How to record product price history

### DISCUSSION THREAD

**Discussion by lukahrobat-lekarnar** [vendurehq/vendure#2341] [Q&A]
Hi, I am wondering what would be the best way to record product price history?

To give a bit more background there is a [EU dircetive](https://eur-lex.europa.eu/eli/dir/2019/2161/oj) that requires online merchants to report the lowest price in the last 30 days if the product is currently discounted to improve consumer information and to enable price comparisons.

The information that needs to be clearly displayed is the original price, the new discounted price and in the last row you can spot the lowest price for this product in the last 30 days. This is an example from aboutyou.com.
![image](https://github.com/vendure-ecommerce/vendure/assets/76471259/6d47d6a5-48da-461a-9a4e-6bc1df6ace28)

Thanks!

**Comment by michaelbromley** [comment]
Hi!

This could be implemented as a plugin. The basic idea would be

1. Define a new entity eg `ProductVariantPriceChange` which stores a price change at a given date for a particular variant.
2. Subscribe to the `ProductVariantEvent` using the EventBus, and check whether the price for the given variant has changed and if so, create a new `ProductVariantPriceChange` entity with the new price and the current date.
3. Define a new field on the ProductVariant type in the Shop API with a resolver which looks up this data for the past 30 days.

This would work well for changes to the actual price. For price changes on account of Promotions, things get considerably more complex due to the dynamic nature of promotions. But I'm not sure whether the directive covers that type of price change.

**Reply by lukahrobat-lekarnar** [reply]
The directive covers only simple single product discounts eg. 10% when purchasing that specific product. Dynamic promotions such as buy this get that, x for 1, free shipping etc. are not covered. Also if the regular price changes are also not in scope.

Thanks for the quick reply! :)

**Reply by michaelbromley** [reply]
> Also if the regular price changes are also not in scope

OK, my answer was specifically covering the regular price changing.

For a Promotion-based price change, the way you connect the price change to the product would vary depending on how your promotion is set up. For instance, let's say you set up a promotion that says "All products with the facet value 'summer-sale' will have a 10% discount", then you can subscribe to the `PromotionEvent` to pick up the creation of that Promotion, but then you'd need to inspect the configured Promotion.conditions to derive the afffected product variants. 

So more complex, but still doable.
